File: cpp/bh_v1_direct.cpp
#include <iostream>
#include <vector>
#include "timer.hpp"
#include <fstream>
#include <cmath>
#include <iomanip>
#include "benchmark_options.hpp"

constexpr float G = 1; // Gravitational constant
constexpr float TIME_STEP = 0.016f; // Time step for the simulation

struct Particle
{
    float velocityX{0.0f};
    float velocityY{0.0f};
    float accX{0.0f};
    float accY{0.0f};
    float mass{0.0f};
    float posX{0.0f};
    float posY{0.0f};
};

struct PhysicsMetrics {
    double totalMomentumX = 0.0;
    double totalMomentumY = 0.0;
    double totalKineticEnergy = 0.0;
    double centerX = 0.0;
    double centerY = 0.0;
};

PhysicsMetrics calculatePhysicsDiagnostics(const std::vector<Particle>& particles) {
    PhysicsMetrics m;
    double totalMass = 0.0;

    for (const auto& p : particles) {
        double mass = static_cast<double>(p.mass);
        double vx = static_cast<double>(p.velocityX);
        double vy = static_cast<double>(p.velocityY);
        double px = static_cast<double>(p.posX);
        double py = static_cast<double>(p.posY);

        m.totalMomentumX += mass * vx;
        m.totalMomentumY += mass * vy;
        m.totalKineticEnergy += 0.5 * mass * (vx * vx + vy * vy);

        m.centerX += mass * px;
        m.centerY += mass * py;
        totalMass += mass;
    }

    m.centerX /= totalMass;
    m.centerY /= totalMass;

    return m;
}

int main(int argc, char** argv)
{
    BenchmarkOptions options;
    if (!parseBenchmarkOptions(argc, argv, options)) return argc > 1 ? 1 : 0;

    Timer timer;
    double totalForceTime = 0.0;
    std::vector<Particle> particles;

    std::ifstream inFile(options.input);
    if (!inFile)
    {
        std::cerr << "Error: Could not open file " << options.input << "!\n";
        return 1;
    }

    Particle p;
    p.accX = 0.0f;
    p.accY = 0.0f;
    
    while (inFile >> p.posX >> p.posY >> p.velocityX >> p.velocityY >> p.mass)
    {
        particles.push_back(p);
    }
    inFile.close();
    if (!validateParticleCount(options, particles.size())) return 1;

    for (int frame = 0; frame < options.totalFrames(); ++frame)
    {
        if (frame == 0) {
            size_t particlesMem = particles.capacity() * sizeof(Particle);
            double totalAppMemMB = static_cast<double>(particlesMem) / (1024.0 * 1024.0);
            std::cout << std::fixed << std::setprecision(6);
            std::cout << "Algorithm memory usage: " << totalAppMemMB << " MB\n";
            std::cout << "Size of Particle: " << sizeof(Particle) << " bytes\n";
        }
        timer.start();
        for (std::size_t i = 0; i < particles.size(); ++i)
        {
            float accX = 0.0f;
            float accY = 0.0f;
            for (std::size_t j = 0; j < particles.size(); ++j)
            {
                if (i == j)
                {
                    continue;
                }

                const auto& particleA = particles[i];
                const auto& particleB = particles[j];
                float dx = particleB.posX - particleA.posX;
                float dy = particleB.posY - particleA.posY;
                float distanceSquared = dx * dx + dy * dy;
                if (distanceSquared < 1e-5f) {
                    continue; 
                }
                float distance = sqrt(distanceSquared);

                float acc = G * particleB.mass / (distanceSquared + 1.0f); // Add small value to prevent division by zero

                accX += acc * (dx / distance);

                accY += acc * (dy / distance);

            }

            particles[i].accX += accX;
            particles[i].accY += accY;
        }
        if (frame == options.warmupFrames && !writeForces(options.dumpForces, particles)) return 1;
        for (auto& particle : particles)
        {
            particle.velocityX += particle.accX * TIME_STEP;
            particle.velocityY += particle.accY * TIME_STEP;
            particle.posX += particle.velocityX * TIME_STEP;
            particle.posY += particle.velocityY * TIME_STEP;
            particle.accX = 0.0f;
            particle.accY = 0.0f;
        }
        const double forceTime = timer.stopTime();
        if (frame >= options.warmupFrames) totalForceTime += forceTime;
        if (frame == options.warmupFrames || frame == options.totalFrames() - 1) {
            auto metrics = calculatePhysicsDiagnostics(particles);
            std::cout << "Frame " << frame << ":\n";
            std::cout << "Momentum (" << metrics.totalMomentumX << ", " << metrics.totalMomentumY << ")\n";
            std::cout << "Kinetic energy " << metrics.totalKineticEnergy << "\n";
            std::cout << "Center of mass (" << metrics.centerX << ", " << metrics.centerY << ")\n";
        }
    }
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Tree construction time: 0.000000 ms / frame\n";
    std::cout << "Force/update time: " << (totalForceTime / options.frames) << " ms / frame\n";
    std::cout << "Cleanup time: 0.000000 ms / frame\n";
    std::cout << "Total measured time: " << totalForceTime << " ms\n";
    return 0;
}

File: nbody_rust/src/bin/bh_v1_direct.rs
use std::time::Instant;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

const TIME_STEP: f32 = 0.016;
const G : f32 = 1.0;

#[path = "../benchmark_options.rs"]
mod benchmark_options;

struct Particle {
    velocity_x: f32,
    velocity_y: f32,
    acc_x: f32,
    acc_y: f32,
    pos_x: f32,
    pos_y: f32,
    mass: f32,
}

pub struct PhysicsMetrics {
    pub total_momentum_x: f64,
    pub total_momentum_y: f64,
    pub total_kinetic_energy: f64,
    pub center_x: f64,
    pub center_y: f64,
}

fn calculate_physics_diagnostics(particles: &[Particle]) -> PhysicsMetrics {
    let mut m = PhysicsMetrics {
        total_momentum_x: 0.0,
        total_momentum_y: 0.0,
        total_kinetic_energy: 0.0,
        center_x: 0.0,
        center_y: 0.0,
    };
    let mut total_mass = 0.0;

    for p in particles {
        // Casting to f64 reduces precision loss
        let mass = p.mass as f64;
        let vx = p.velocity_x as f64;
        let vy = p.velocity_y as f64;
        let px = p.pos_x as f64;
        let py = p.pos_y as f64;

        m.total_momentum_x += mass * vx;
        m.total_momentum_y += mass * vy;
        m.total_kinetic_energy += 0.5 * mass * (vx * vx + vy * vy);

        m.center_x += mass * px;
        m.center_y += mass * py;
        total_mass += mass;
    }

    m.center_x /= total_mass;
    m.center_y /= total_mass;

    m
}

fn main() {
    let options = match benchmark_options::parse_options() {
        Ok(options) => options,
        Err(message) => { eprintln!("{message}"); std::process::exit(1); }
    };
    let mut particles = Vec::new();

    let path = Path::new(&options.input);
    let file = match File::open(&path) {
        Ok(f) => f,
        Err(_) => {
            eprintln!("Error: Could not open input file {}.", options.input);
            std::process::exit(1);
        }
    };

    let reader = io::BufReader::new(file);

    for line in reader.lines() {
        let line = line.expect("Error while reading a line from the file");

        // Split the line by spaces
        let parts: Vec<&str> = line.split_whitespace().collect();

        if parts.len() == 5 {
            particles.push(Particle {
                pos_x: parts[0].parse().unwrap(),
                pos_y: parts[1].parse().unwrap(),
                velocity_x: parts[2].parse().unwrap(),
                velocity_y: parts[3].parse().unwrap(),
                mass: parts[4].parse().unwrap(),
                acc_x: 0.0,
                acc_y: 0.0,
            });
        }
    }
    if let Err(message) = benchmark_options::validate_particle_count(&options, particles.len()) {
        eprintln!("{message}");
        std::process::exit(1);
    }

    let mut total_force_time_ms = 0.0;

    for frame in 0..options.total_frames() {

        let start_time = Instant::now();

        for i in 0..particles.len() {
            let mut acc_x: f32 = 0.0;
            let mut acc_y: f32 = 0.0;
            let particle_a_pos_x = particles[i].pos_x;
            let particle_a_pos_y = particles[i].pos_y;

            for j in 0..particles.len() {
                if i == j {
                    continue;
                }
                let particle_b = &particles[j];
                let dx = particle_b.pos_x - particle_a_pos_x;
                let dy = particle_b.pos_y - particle_a_pos_y;
                let dist_sq = dx * dx + dy * dy;
                if dist_sq < 1e-5 {
                    continue;
                }

                let distance = dist_sq.sqrt();

                let acc = G * particle_b.mass / (dist_sq + 1.0);

                acc_x += acc * (dx / distance);
                acc_y += acc * (dy / distance);
            }
            particles[i].acc_x = acc_x;
            particles[i].acc_y = acc_y;
        }

        if frame == options.warmup_frames {
            if let Some(path) = options.dump_forces.as_deref() {
                if let Err(message) = benchmark_options::write_forces(
                    path,
                    particles.iter().map(|particle| (particle.acc_x, particle.acc_y)),
                ) {
                    eprintln!("{message}");
                    return;
                }
            }
        }

        // Euler integration (position update)
        for i in 0..particles.len() {
            particles[i].velocity_x += particles[i].acc_x * TIME_STEP;
            particles[i].velocity_y += particles[i].acc_y * TIME_STEP;
            particles[i].pos_x += particles[i].velocity_x * TIME_STEP;
            particles[i].pos_y += particles[i].velocity_y * TIME_STEP;

            particles[i].acc_x = 0.0;
            particles[i].acc_y = 0.0;
        }
        let force_time_ms = start_time.elapsed().as_secs_f64() * 1000.0;
        if frame >= options.warmup_frames {
            total_force_time_ms += force_time_ms;
        }

        if frame == options.warmup_frames || frame == options.total_frames() - 1 {
            let metrics = calculate_physics_diagnostics(&particles);
            println!("Frame {}:", frame);
            println!("Momentum ({:.6}, {:.6})", metrics.total_momentum_x, metrics.total_momentum_y);
            println!("Kinetic energy {:.6}", metrics.total_kinetic_energy);
            println!("Center of mass ({:.6}, {:.6})", metrics.center_x, metrics.center_y);
        }
    }

    println!("Tree construction time: 0.0000 ms / frame");
    println!("Force/update time: {:.4} ms / frame", total_force_time_ms / (options.frames as f64));
    println!("Cleanup time: 0.0000 ms / frame");
    println!("Total measured time: {:.4} ms", total_force_time_ms);
}
